src/toil/batchSystems/singleMachine.py

```diff
@@ -302,38 +303,69 @@ def _stop_now(self, popens: Sequence[subprocess.Popen]) -> List[int]:
 
-            logger.debug('Process %s known to batch system %s is stopped; it returned %s', popen.pid, id(self), popen.returncode)
+        # Make sure all child processes have received their kill signal
+        self._wait_for_death(pgids, timeout)
 
-        for pgid in pgids:
-            try:
-                # Send a kill to the group again, to see if anything in it
-                # is still alive. Our first kill might not have been
-                # delivered yet.
-                os.killpg(pgid, signal.SIGKILL)
-                # If that worked it is still alive, let user know that we may leave
-                # behind dead but unreaped processes.
-                logger.warning('Sent redundant shutdown kill to surviving process group %s known to batch system %s', pgid, id(self))
-                logger.warning('Make sure your jobs are cleaning up child processes appropriately to avoid zombie '
-                               'processes possibly being left behind.')
+    def _wait_for_death(self, pgids: List[int], timeout: int = 5):
+        """
+        Wait for the process groups to be killed. Blocks until the processes
+        are gone or timeout is passed.
 
-            except ProcessLookupError:
-                # The group is actually gone now.
-                pass
-            except PermissionError:
-                # The group is not only gone but reused
-                pass
+        :param pgids: The list of process group ids.
+        :param timeout: The number of seconds to wait for all process groups to
+                        be gone.
+        """
+        # TODO: this opens a PGID reuse risk; someone else might've reaped the
+        #  process and its PGID may have been re-used.
+
+        start = datetime.datetime.now()
+        while len(pgids) > 0 and (datetime.datetime.now() - start).total_seconds() < timeout:
+            new_pgids: List[int] = []
+            for pgid in pgids:
+                try:
+                    # Send a kill to the group again, to see if anything in it
+                    # is still alive. Our first kill might not have been
+                    # delivered yet.
+                    os.killpg(pgid, signal.SIGKILL)
+
+                    # If we reach here, something in the process group still
+                    # exists.
+                    new_pgids.append(pgid)
+                except ProcessLookupError:
+                    # The group is actually gone now.
+                    pass
+                except PermissionError:
+                    # The group is not only gone but reused
+                    pass
+
+            pgids = new_pgids
+            if len(pgids) > 0:
+                time.sleep(0.1)
+
+        if len(pgids) > 0:
+            # If any processes are still alive, let user know that we may leave
+            # behind dead but unreaped processes.
+            logger.warning('Process groups %s were not reaped.', str(pgids))
```

Comment:
You don't really reap a process group, just a process. Maybe:

```suggestion
            logger.warning('Processes were not reaped in groups: %s', str(pgids))
```